clients/src/main/java/org/apache/kafka/common/record/Compressor.java

```diff
@@ -271,4 +293,33 @@ static public DataInputStream wrapForInput(ByteBufferInputStream buffer, Compres
             throw new KafkaException(e);
         }
     }
+
+    private interface ConstructorSupplier {
+        Constructor get() throws ClassNotFoundException, NoSuchMethodException;
+    }
+
+    // this code is based on Guava's @see{com.google.common.base.Suppliers.MemoizingSupplier}
+    private static class MemoizingConstructorSupplier {
+        final ConstructorSupplier delegate;
+        transient volatile boolean initialized;
+        transient Constructor value;
```

Comment:
Why not replace the `initialized` check with a null check? In that case, `value` needs to be volatile and `initialized` can be removed.